staging/src/k8s.io/client-go/tools/cache/delta_fifo.go

```diff
@@ -139,20 +139,18 @@ type DeltaFIFO struct {
 }
 
 // TransformFunc allows for transforming an object before it will be processed.
-// TransformFunc (similarly to ResourceEventHandler functions) should be able
-// to correctly handle the tombstone of type cache.DeletedFinalStateUnknown.
-//
-// New in v1.27: In such cases, the contained object will already have gone
-// through the transform object separately (when it was added / updated prior
-// to the delete), so the TransformFunc can likely safely ignore such objects
-// (i.e., just return the input object).
+// TransformFunc (similarly to ResourceEventHandler functions).
 //
 // The most common usage pattern is to clean-up some parts of the object to
 // reduce component memory usage if a given component doesn't care about them.
 //
-// New in v1.27: unless the object is a DeletedFinalStateUnknown, TransformFunc
-// sees the object before any other actor, and it is now safe to mutate the
-// object in place instead of making a copy.
+// New in v1.27: TransformFunc sees the object before any other actor, and it
+// is now safe to mutate the object in place instead of making a copy.
+//
+// It's recommended for the TransformFunc to be idempotent.
+// It MUST be idempotent if objects already present in the cache are passed to
+// the Replace() to avoid re-mutating them. Default informers do not pass
+// existing objects to Replace though.
```

Comment:
I'm a little wary of this constraint. Consumers of the `TransformFunc` will not be aware of nor have control over how or where the object they receive came from. Nor should they care. Also the `DeltaFIFO` does not have control over the origin of the objects from the upstream informers. I think it would be ideal for the `DeltaFIFO` to provide protection for the `TransformFunc` from potential data races. If an object already exists in the `knownObjects` cache and, more specifically, points to the same **instance** then make a copy of it before passing to the `TransformFunc`. The `TransformFunc` would then truly be free to mutate the object w/o any additional idempotency constraint.

Reply:
I agree that preventing more races is desired, however these for now these don't expose default usage of it. We should protect it further, but preventing existing issues seems more important to start with.

Reply:
> however these for now these don't expose default usage of it.

I'm not clear on what you mean by this.

I think the text above would be a bit vague and confusing to users of the API. It would be for me if I hadn't studied the code and understand the issue and what it's trying say. With the changes in this PR, the potential race issue with passing cached objects to the `DeltaFIFO` could only occur with a custom informer and it's probably unlikely anyone is doing that so I'd omit this text. In fact, this should really documented as part of the `DeltaFIFO` API.

Reply:
>> however these for now these don't expose default usage of it.
>
> I'm not clear on what you mean by this.

After the fix in this PR to omit Sync events and DeletedFinalStateUnknown objects, none of the uses of DeltaFIFO that client-go sets up (default informers / clients, etc) feed objects already present in `knownObjects` back into the transformer.

That means the only way a DeltaFIFO user would encounter that is if they wired up a non-default usage of DeltaFIFO to pass the same objects into DeltaFIFO multiple times as if they were new objects.

Someone doing that is presumably the same person wiring up the TransformFunc, and they need to understand the implications of what they are doing. Doing this effectively already breaks the *TransformFunc sees the object before any other actor, and it is now safe to mutate the object in place instead of making a copy.* statement / guarantee, so the TransformFunc author _must_ be aware of the implications of mutating objects it gets when they are being provided by something with different behavior than default client-go clients / informers.

> If an object already exists in the `knownObjects` cache and, more specifically, points to the same **instance** then make a copy of it before passing to the `TransformFunc`

The tracking / comparison cost to safeguard `TransformFunc` in that case is too high, and even if we did, `TransformFunc` would have to be aware of and protect against disrupting a non-default object provider.

Reply:
I understand the reasoning. My point is that the `TransformFunc` is part of a higher-level API and is much more widely used and visible then the `DeltaFIFO`. The `New*Informer` functions are really the external-facing aspect of the `cache` module and I'm sure most users of these aren't even aware of `DeltaFIFO` and its semantics. As you mentioned, the author of a non-default usage of `DeltaFIFO`, if there are any, is "_presumably the same person wiring up the TransformFunc_", therefore document the constraints and implications re: caching and the `TransformFunc` in the `DeltaFIFO` API rather than here. As I said, I think the paragraph above will only confuse most people. Anyway, that's my 2 cents...